File: Reverse_linked_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
//recursively version
 /*   ListNode* reverseList(ListNode* head) {
        if(head == NULL)
            return NULL;
        if(head->next == NULL)
            return head;
        ListNode *head1 = NULL, *cur = NULL;
        
        head1 = reverseList(head->next);
        cur = head1;
        while(cur->next)
        {
            cur = cur->next;
        }
        cur->next = head;
        head->next = NULL;
        
        return head1;
    }
*/
//iteratively version
    ListNode *reverseList(ListNode* head){
        ListNode *first = NULL, *cur = NULL;
        if(!head)
            return NULL;
        first = head;
        cur = head->next;
        head->next = NULL;
        while(cur)
        {
            ListNode *cur1 = cur->next;
            cur->next = first;
            first = cur;
            cur = cur1;
        }
        
        return first;
    }
 };

File: lintcode/LinkedList/reverse_linked_list.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
	if head == nil{
		return head
	}
	h, e := helper(head)
	e.Next = nil
	return h
}

func helper(head *ListNode) (*ListNode, *ListNode){
	if head == nil || head.Next == nil{
		return head, head
	}
	h, e := helper(head.Next)
	e.Next = head
	return h, head
}

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
// no recursion
//func reverseList(head *ListNode) *ListNode {
//	if head == nil || head.Next == nil{
//		return head
//	}
//	pre, cur := head, head.Next
//	pre.Next = nil
//	n := ListNode{0, head}
//
//	for cur != nil{
//		tmp := cur.Next
//		cur.Next = pre
//		pre = cur
//
//		n.Next = pre
//		cur = tmp
//	}
//	return n.Next
//}
